File: Plugins/ScriptPlugins/ActionOnReport.js
const init = (registerEventCallback, serviceResolver, configWrapper) => {
    plugin.onLoad(serviceResolver, configWrapper);

    registerEventCallback('IManagementEventSubscriptions.ClientPenaltyAdministered', (penaltyEvent, _) => {
        plugin.onPenalty(penaltyEvent);
    });

    return plugin;
};

const plugin = {
    author: 'RaidMax',
    version: '2.1',
    name: 'Action on Report',
    config: {
        enabled: false, // indicates if the plugin is enabled
        reportAction: 'TempBan', // can be TempBan or Ban
        maxReportCount: 5, // how many reports before action is taken
        tempBanDurationMinutes: 60 // how long to temporarily ban the player
    },

    onPenalty: function (penaltyEvent) {
        if (!this.config.enabled || penaltyEvent.penalty.type !== 'Report') {
            return;
        }
        
        if (!penaltyEvent.client.isIngame || (penaltyEvent.client.level !== 'User' && penaltyEvent.client.level !== 'Flagged')) {
            this.logger.logInformation(`Ignoring report for client (id) ${penaltyEvent.client.clientId} because they are privileged or not in-game`);
            return;
        }

        let reportCount = this.reportCounts[penaltyEvent.client.networkId] === undefined ? 0 : this.reportCounts[penaltyEvent.Client.NetworkId];
        reportCount++;
        this.reportCounts[penaltyEvent.client.networkId] = reportCount;

        if (reportCount >= this.config.maxReportCount) {
            switch (this.config.reportAction) {
                case 'TempBan':
                    this.logger.logInformation(`TempBanning client (id) ${penaltyEvent.client.clientId} because they received ${reportCount} reports`);
                    penaltyEvent.client.tempBan(this.translations['PLUGINS_REPORT_ACTION'], System.TimeSpan.FromMinutes(this.config.tempBanDurationMinutes), penaltyEvent.Client.CurrentServer.asConsoleClient());
                    break;
                case 'Ban':
                    this.logger.logInformation(`Banning client (id) ${penaltyEvent.client.clientId} because they received ${reportCount} reports`);
                    penaltyEvent.client.ban(this.translations['PLUGINS_REPORT_ACTION'], penaltyEvent.client.currentServer.asConsoleClient(), false);
                    break;
            }
        }
    },

    onLoad: function (serviceResolver, configWrapper) {
        this.translations = serviceResolver.resolveService('ITranslationLookup');
        this.logger = serviceResolver.resolveService('ILogger', ['ScriptPluginV2']);
        this.configWrapper = configWrapper;

        const storedConfig = this.configWrapper.getValue('config', newConfig => {
            if (newConfig) {
                plugin.logger.logInformation('ActionOnReport config reloaded. Enabled={Enabled}', newConfig.enabled);
                plugin.config = newConfig;
            }
        });

        if (storedConfig != null) {
            this.config = storedConfig
        } else {
            this.configWrapper.setValue('config', this.config);
        }
        
        this.logger.logInformation('ActionOnReport {version} by {author} loaded. Enabled={Enabled}', this.version, this.author, this.config.enabled);
        this.reportCounts = {};
    }
};

File: Plugins/ScriptPlugins/ActionOnReport.cs
#:package RaidMax.IW4MAdmin.SharedLibraryCore@2026.1.6.1

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Data.Models.Client;
using Microsoft.Extensions.Logging;
using SharedLibraryCore;
using SharedLibraryCore.Events.Management;
using SharedLibraryCore.Interfaces;
using SharedLibraryCore.Interfaces.Events;
using Microsoft.Extensions.DependencyInjection;

/// <summary>
/// Action on Report Plugin - Automatically bans or temporarily bans players
/// after they receive a certain number of reports.
/// </summary>
public class ActionOnReportPlugin : IPluginV2
{
    public static void RegisterDependencies(IServiceCollection serviceCollection)
    {
        serviceCollection.AddConfiguration<ActionOnReportConfig>(
            "ActionOnReportSettings",
            new ActionOnReportConfig());
    }

    public string Name => "Action on Report";
    public string Author => "RaidMax";
    public string Version => "2.1";

    private readonly ILogger<ActionOnReportPlugin> _logger;
    private readonly ActionOnReportConfig _config;
    private readonly ITranslationLookup _translationLookup;
    private readonly Dictionary<long, int> _reportCounts = new();

    public ActionOnReportPlugin(
        ILogger<ActionOnReportPlugin> logger,
        ActionOnReportConfig config,
        ITranslationLookup translationLookup)
    {
        _logger = logger;
        _config = config;
        _translationLookup = translationLookup;

        // Subscribe to penalty events
        IManagementEventSubscriptions.ClientPenaltyAdministered += OnPenalty;

        _logger.LogInformation("ActionOnReport {Version} by {Author} loaded. Enabled={Enabled}", 
            Version, Author, _config.Enabled);
    }


    private Task OnPenalty(ClientPenaltyEvent penaltyEvent, CancellationToken token)
    {
        if (!_config.Enabled || penaltyEvent.Penalty.Type != Data.Models.EFPenalty.PenaltyType.Report)
        {
            return Task.CompletedTask;
        }

        var client = penaltyEvent.Client;

        // Ignore if client is not in-game or is privileged
        if (!client.IsIngame || (client.Level != EFClient.Permission.User && client.Level != EFClient.Permission.Flagged))
        {
            _logger.LogInformation("Ignoring report for client (id) {ClientId} because they are privileged or not in-game", 
                client.ClientId);
            return Task.CompletedTask;
        }

        // Get and increment report count
        if (!_reportCounts.TryGetValue(client.NetworkId, out var reportCount))
        {
            reportCount = 0;
        }
        reportCount++;
        _reportCounts[client.NetworkId] = reportCount;

        if (reportCount >= _config.MaxReportCount)
        {
            var reason = _translationLookup["PLUGINS_REPORT_ACTION"] ?? "Too many reports";

            switch (_config.ReportAction)
            {
                case "TempBan":
                    _logger.LogInformation("TempBanning client (id) {ClientId} because they received {ReportCount} reports", 
                        client.ClientId, reportCount);
                    client.TempBan(reason, TimeSpan.FromMinutes(_config.TempBanDurationMinutes), 
                        client.CurrentServer.AsConsoleClient());
                    break;

                case "Ban":
                    _logger.LogInformation("Banning client (id) {ClientId} because they received {ReportCount} reports", 
                        client.ClientId, reportCount);
                    client.Ban(reason, client.CurrentServer.AsConsoleClient(), false);
                    break;
            }
        }

        return Task.CompletedTask;
    }

    public void Dispose()
    {
        IManagementEventSubscriptions.ClientPenaltyAdministered -= OnPenalty;
        _logger.LogInformation("ActionOnReport unloaded");
    }
}

/// <summary>
/// Configuration class for ActionOnReport plugin.
/// </summary>
public class ActionOnReportConfig
{
    /// <summary>
    /// Indicates if the plugin is enabled.
    /// </summary>
    public bool Enabled { get; set; } = false;

    /// <summary>
    /// Action to take when report threshold is reached. Can be "TempBan" or "Ban".
    /// </summary>
    public string ReportAction { get; set; } = "TempBan";

    /// <summary>
    /// How many reports before action is taken.
    /// </summary>
    public int MaxReportCount { get; set; } = 5;

    /// <summary>
    /// How long to temporarily ban the player (in minutes).
    /// </summary>
    public int TempBanDurationMinutes { get; set; } = 60;
}
